engine/include/Engine/Intermediate/GameObject.inl

```diff
@@ -0,0 +1,179 @@
+#include "Engine/Intermediate/GameObject.hpp"
+
+GameObject::GameObject(const CreateArg& arg)
+    : m_name{arg.name}, m_pTransform{&DataChunk<TransformComponent>::getInstance()->addComponent(*this,
+                                                                                                 arg.transformArg)},
+      m_pComponents{}
+{
+}
+
+GameObject::GameObject()
+    : m_name{""}, m_pTransform{&DataChunk<TransformComponent>::getInstance()->addComponent(*this)}, m_pComponents{}
+{
+}
+
+template <typename T>
+void GameObject::updateComponentLink(const T* oldPtr, T* newPtr) noexcept
+{
+    GPE_ASSERT(oldPtr != nullptr, "oldPtr cannot be nullptr");
+    GPE_ASSERT(newPtr != nullptr, "newPtr cannot be nullptr");
+
+    for (Component* pComponent : m_pComponents)
+    {
+        T* checkedCompPtr = dynamic_cast<T*>(pComponent);
+
+        if (checkedCompPtr == oldPtr)
+            [[unlikely]]
+            {
+                pComponent = newPtr;
+                return;
+            }
+    };
+}
+
+template <typename T, typename... Args>
+T& GameObject::addComponent(Args&&... args) noexcept
+{
+    return m_pComponents.emplace_back(&DataChunk<T>::getInstance()->addComponent(*this, std::forward<Args>(args)...));
+}
+
+template <typename T>
+T* GameObject::getComponent() noexcept
+{
+    for (auto&& pComponent : m_pComponents)
+    {
+        T* comp = dynamic_cast<T*>(pComponent);
```

Comment:
Is there a class derived from T that you would like to get from getComponent() ?
If no, using typeid + static_cast is faster.